### PR TITLE
(#251): Surface::create_similar() and friends should return a Result

### PR DESCRIPTION
**Pull request by federicomenaquintero**
Cairo's surface creation functions never return NULL; instead they
always return a surface, but it may be in an error state.  In #141 we
started making the binding functions return Result for this; some
returned the plain FooSurface type, some others Option<FooSurface>.

This makes the following functions return Result<FooSurface, Status>

  Surface::create_similar()
  Surface::create_similar_image()
  RecordingSurface::create()
  Device.surface_create()
  Device.surface_create_for_target()

The foundation for all of this is that Surface::from_raw_full() now
also returns Result<FooSurface, Status>.  This is to make things
consistent with ImageSurface::from_raw_full().  Analogously, we now
have RecordingSurface::from_raw_full() that also returns Result.

Fixes https://github.com/gtk-rs/cairo/issues/251

### CODE REVIEW

**Comment by federicomenaquintero** [comment]
Bleh, forgot to check  the build with features.  I'll fix this tomorrow.

**Comment by federicomenaquintero** [comment]
Fixed the features.

Also, went along and fixed all the PdfSurface functions that may set &self in an error state.

**Comment by EPashkin** [comment]
AppVeyor's errors are legit

File: src/pdf.rs [hunk 1]
```diff
@@ -170,7 +176,7 @@ impl FromGlibPtrFull<*mut ffi::cairo_surface_t> for PdfSurface {
     #[inline]
     unsafe fn from_glib_full(ptr: *mut ffi::cairo_surface_t) -> PdfSurface {
         Self {
-            inner: Surface::from_raw_full(ptr),
+            inner: Surface::from_raw_full(ptr).unwrap(),
```
[the remark this answers is not in the document]
Comment (federicomenaquintero) [reply src/pdf.rs]: Uuh, I think this needs a treatment similar to ImageSurface, with TryFrom.

File: src/pdf.rs [hunk 2]
```diff
@@ -60,20 +59,22 @@ impl PdfSurface {
         vers_slice.iter().map(|v| PdfVersion::from(*v))
     }
 
-    pub fn restrict(&self, version: PdfVersion) {
+    pub fn restrict(&self, version: PdfVersion) -> Result<(), Status> {
         unsafe {
             ffi::cairo_pdf_surface_restrict_to_version(self.inner.to_raw_none(), version.into());
         }
+        self.status().to_result(())
```
Comment (sdroege) [inline src/pdf.rs]: What does this mean exactly? Is the surface still usable afterwards or is it completely broken? :)

**Comment by sdroege** [comment]
Looks good to me generally, thanks for the work :)

File: src/pdf.rs (hunk 2 above)
Comment (federicomenaquintero) [reply src/pdf.rs]: Most cairo_pdf_surface_*() functions start by calling [`_extract_pdf_surface()`](https://gitlab.freedesktop.org/cairo/cairo/blob/52a7c79f/src/cairo-pdf-surface.c#L620-665), which can set the surface in an error state.

Afterwards... not all of cairo_surface_*() check for the surface->status.  It's not completely unusable if in error, but some functions *will* still work ¯\_(ツ)_/¯

**Comment by federicomenaquintero** [comment]
Let me see if I can fix the TryFrom thing - to make it consistent with other concrete surfaces that wrap the abstract one.

**Comment by sdroege** [comment]
Any updates here?

**Comment by sdroege** [comment]
@federicomenaquintero we'd like to get a release out next weekend. Any chance this can be ready by then? :)

**Comment by federicomenaquintero** [comment]
I'll try to work on this this week.  Need to remind myself what's left to be done :)

**Comment by sdroege** [comment]
@federicomenaquintero Thanks, let us know when it's time for another review or if you're stuck somewhere :)

**Comment by federicomenaquintero** [comment]
I think it's ready.

The next step is probably to take all the functions that now all look alike - FooSurface::try_from(Surface); the to_glib ones - and make a macro for them or something.

**Comment by federicomenaquintero** [comment]
Argh, QuartzSurface. I'll deal with it.

**Comment by federicomenaquintero** [comment]
Now all derived surfaces are done with a macro.  This even buys us some methods that were missing from the old implementations.

**Comment by sdroege** [comment]
Looks good to me, thanks! This cleans up everything quite nicely.

The CI failure in the examples is expected, so this is IMHO good to go once all jobs have failed only when building the examples and not before that @GuillaumeGomez

**Comment by sdroege** [comment]
@GuillaumeGomez Please trigger CI once https://github.com/gtk-rs/examples/pull/271 is merged and then this can go in IMHO

**Comment by sdroege** [comment]
@GuillaumeGomez @federicomenaquintero minimum rust version in the CI needs to be updated to 1.39 or it will fail :) Can one of you push a fix for that to this branch?

**Comment by sdroege** [comment]
Now only misses a `cargo fmt` run. @GuillaumeGomez? :)

**Comment by federicomenaquintero** [comment]
Uuuuh, what changed that requires Rust 1.39?  I mean, I need to update things in Suse, anyway, but just to know.

**Comment by federicomenaquintero** [comment]
Thanks for the merge!

**Comment by GuillaumeGomez** [comment]
async/await in other crates. That wouldn't make sense to have cairo handle its version on its own.

**Comment by federicomenaquintero** [comment]
Got it.  Sounds good, thanks!

**Pull request merged**